lib/shared/services/request_cache_service.dart

```diff
@@ -0,0 +1,102 @@
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:collection/collection.dart';
+import 'package:hive/hive.dart';
+import 'package:http/http.dart' as http;
+
+class RequestCacheService<T> {
+  final T Function(Map<String, dynamic>) fromJson;
+  final Map<String, dynamic> Function(T) toJson;
+  final Duration cacheDuration;
+  final String networkCacheKey;
+  final http.Client httpClient;
+
+  RequestCacheService({
+    required this.fromJson,
+    required this.toJson,
+    this.cacheDuration = const Duration(minutes: 1),
+    this.networkCacheKey = 'network_cache',
+    http.Client? httpClient,
+  }) : httpClient = httpClient ?? http.Client();
+
+  /// Fetches data from the cache and API.
+  Stream<T> fetchData(String url) async* {
+    final box = await Hive.openBox(networkCacheKey);
```

Comment:
instead of opening and closing `Box` whenever `fetchData()` is called, we can just have `RequestCacheService` take in a `Box` (take a look at `PfisNotifier`, `VcsNotifier` for some examples)